BDF-Chainlit-Chatbot/PDC_tools.py

```diff
@@ -0,0 +1,582 @@
+from typing import List
+#from llama_index.llms.openai import OpenAI
+import json, requests
+from io import StringIO
+from typing import List
+import pandas as pd;
+
+# This is a set of helper function for retrieving and processing data from PDC
+
+pdc_url = 'https://proteomic.datacommons.cancer.gov/graphql?query='
+
+# Helper function to cache all diseases and primary sites from PDC
+# TO DO: Use a proper metadata source for these values e.g. PDC Dictionary or caDSR
+def getAllDiseasesAndPrimarySites():
+    all_studies = []
+    all_studies_map = {}
+    print('Caching studies')
+    # Cache all studies
+    url = pdc_url
+    url += '''{allPrograms 
+                   { 
+                    projects  
+                      {  
+                        studies  
+                          { pdc_study_id study_submitter_id submitter_id_name 
+                            analytical_fraction study_name disease_types 
+                            primary_sites  experiment_type acquisition_type
+                          } 
+                        }
+                    }
+                }'''
+    response = requests.get(url)
+    if(response.ok):
+        study_list = json.loads(response.content)
+        study_list = study_list['data']['allPrograms']
+        if (study_list is not None):
+            for programs in study_list:
+                for projects in study_list:
+                    for project in projects['projects']:
+                        for study in project['studies']:
+                            all_studies.append(study)
+                            all_studies_map[study['pdc_study_id']] = study
+    url = pdc_url + '''{allPrograms {program_id  
+                program_submitter_id  name 
+                projects  {
+                    project_id  project_submitter_id  name  
+                    studies  {
+                        pdc_study_id study_id study_submitter_id submitter_id_name 
+                        analytical_fraction study_name disease_types 
+                        primary_sites  
+                        experiment_type acquisition_type} }}}'''
+    response = requests.get(url)
+    if(response.ok):
+        #If the response was OK then print the returned JSON
+        jData = json.loads(response.content)
+        jData = jData['data']['allPrograms']
+        return all_studies, jData
+    else:
+      # If response code is not ok (200), print the resulting http error code with description
+      response.raise_for_status()
+        
+
+# This is a helper function
+def getDiseaseInformation(mdata:object)->str:
+    '''
```

Comment:
some comments are before the function using # and others imbedded using ''', should probably come up with a style guide and be consistent.